File: src/main/python/geeksforgeeks/arrays/accepted-in-other-languages/Rearrange Array Alternately.cpp
// C++ program to rearrange an array in minimum
// maximum form
#include <bits/stdc++.h>
using namespace std;

// Prints max at first position, min at second position
// second max at third position, second min at fourth
// position and so on.
void rearrange(long arr[], long n)
{
	// initialize index of first minimum and first
	// maximum element
	long max_idx = n - 1, min_idx = 0;

	// store maximum element of array
	long max_elem = arr[n - 1] + 1;

	// traverse array elements
	for (long i = 0; i < n; i++) {
		// at even index : we have to put maximum element
		if (i % 2 == 0) {
			arr[i] += (arr[max_idx] % max_elem) * max_elem;
			max_idx--;
		}

		// at odd index : we have to put minimum element
		else {
			arr[i] += (arr[min_idx] % max_elem) * max_elem;
			min_idx++;
		}
	}

	// array elements back to it's original form
	for (long i = 0; i < n; i++)
		arr[i] = arr[i] / max_elem;
}

// Driver program to test above function
int main()
{
	long t, n, *arr;
	cin>>t;
	while(t--) {
	    cin>>n;
	    arr = new(nothrow) long[n];
	    for(long i=0; i<n; i++) {
	        cin>>arr[i];
	    }
	    rearrange(arr, n);
	    for(long i=0; i<n; i++) {
	        cout<<arr[i]<<' ';
	    }
	    cout<<endl;
	    delete arr;
	}
}

File: src/main/python/geeksforgeeks/arrays/accepted-in-other-languages/Rearrange Array Alternately.py
"""
https://practice.geeksforgeeks.org/problems/-rearrange-array-alternately/0/
https://www.geeksforgeeks.org/rearrange-array-maximum-minimum-form/
https://www.geeksforgeeks.org/rearrange-array-maximum-minimum-form-set-2-o1-extra-space/
"""


# Prints max at first position, min at second position
# second max at third position, second min at fourth position and so on
# Assumption: The provided array is already sorted in asc order!!
def rearrangeWithNSpace(arr, n):
    print('This solution should give Memory Error in GeeksForGeeks Practise,'
          ' since we are not supposed to use more than O(1) space')
    auxiliary_array = [None] * n
    small, large = 0, n-1
    for i in range(n):
        if i %2 is 0:
            auxiliary_array[i] = arr[large]
            large -= 1
        else:
            auxiliary_array[i] = arr[small]
            small += 1

    for i in auxiliary_array:
        print(i, end=' ')
    print()
    # return auxiliary_array;


# We will basically do what the package name boasts about :P
def rearrangeWithConstantExtraSpace(arr, n):
    # Get the value of biggest dude amongst them all,
    # which is basically any number gt arr[n-1] since arr is sorted in asc order
    the_big_dude = arr[n-1] + 1

    # We still need to iterate from both the ends, so we will keep 2 var, and proceed like the previous case.
    small, large = 0, n-1

    # But unlike last case, we will store the other dude at same index in same array
    for i in range(n):
        if i%2 is 0:
            arr[i] += (arr[large] % the_big_dude) * the_big_dude
            large -= 1
        else:
            arr[i] += (arr[small] % the_big_dude) * the_big_dude
            small += 1

    # Now that the parasite is home, let's kill the host
    for i in range(n):
        arr[i] //= the_big_dude

    for i in arr:
        print(i, end=' ')
    print()
    # return arr


if __name__ == '__main__':
    t = int(input())
    while t > 0:
        n = int(input())
        arr = [int(x) for x in input().split()]
        rearrangeWithNSpace(arr, n)
        rearrangeWithConstantExtraSpace(arr, n)
        t -= 1
